modules/pipeline/providers/leaderworker/worker.go

```diff
@@ -0,0 +1,390 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leaderworker
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/pkg/errors"
+
+	"github.com/erda-project/erda/modules/pipeline/providers/leaderworker/worker"
+	"github.com/erda-project/erda/pkg/strutil"
+)
+
+func (p *provider) listWorkersByType(ctx context.Context, typ worker.Type) ([]worker.Worker, error) {
+	getResp, err := p.EtcdClient.Get(ctx, p.makeEtcdWorkerKeyPrefix(typ), clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+	var workers []worker.Worker
+	for _, kv := range getResp.Kvs {
+		w, err := worker.NewFromBytes(kv.Value)
+		if err != nil {
+			return nil, err
+		}
+		w.SetType(typ)
+		workers = append(workers, w)
+	}
+	return workers, nil
+}
+
+func (p *provider) getWorkerTaskLogicIDFromIncomingKey(workerID worker.ID, key string) string {
+	prefix := p.makeEtcdWorkerLogicTaskListenPrefix(workerID)
+	return strutil.TrimPrefixes(key, prefix)
+}
+
+// see: makeEtcdWorkerTaskDispatchKey
+func (p *provider) getWorkerIDFromIncomingKey(key string) worker.ID {
+	prefix := p.makeEtcdWorkerGeneralDispatchPrefix()
+	if !strutil.HasPrefixes(key, prefix) {
+		return ""
+	}
+	workerIDAndSuffix := strutil.TrimPrefixes(key, prefix)
+	workerIDAndLogicTaskID := strutil.Split(workerIDAndSuffix, "/task/")
+	if len(workerIDAndLogicTaskID) != 2 {
+		return ""
+	}
+	return worker.ID(workerIDAndLogicTaskID[0])
+}
+
+func (p *provider) getWorkerLastHeartbeatUnixTime(ctx context.Context, workerID worker.ID) (lastHeartbeatUnitTime int64, err error, needRetry bool) {
+	getResp, err := p.EtcdClient.Get(ctx, p.makeEtcdWorkerHeartbeatKey(workerID))
+	if err != nil {
+		return 0, err, true
+	}
+	if len(getResp.Kvs) == 0 {
+		return 0, fmt.Errorf("no heartbeat info"), false
+	}
+	lastHeartbeatTime, err := strconv.ParseInt(string(getResp.Kvs[0].Value), 10, 64)
+	if err != nil {
+		return 0, err, false
+	}
+	return lastHeartbeatTime, nil, false
+}
+
+func (p *provider) getWorkerFromEtcd(ctx context.Context, workerID worker.ID, typ worker.Type) (worker.Worker, error) {
+	key := p.makeEtcdWorkerKey(workerID, typ)
+	getResp, err := p.EtcdClient.Get(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
+	if getResp == nil || getResp.Count == 0 || len(getResp.Kvs) == 0 {
+		return nil, worker.ErrNotFound
+	}
+	kv := getResp.Kvs[0]
+	if kv == nil || len(kv.Value) == 0 {
+		return nil, worker.ErrNotFound
+	}
+	w, err := worker.NewFromBytes(kv.Value)
+	if err != nil {
+		return nil, err
+	}
+	w.SetType(typ)
+	return w, nil
+}
+
+func (p *provider) checkWorker(ctx context.Context, w worker.Worker) error {
```

Comment:
is `validateWorker` better? or `workerValidation`

Reply:
no

Reply:
**DO NOT SPLIT HAIRS.**

these two words are similar here.

You can specify it in Erda-Go-Specification please. And then I will follow the rules.

Reply:
then, at least `checkWorkerIsReady`

Reply:
nice, accepted.